Week of Nov 6 edits (#61)
xregistry/spec: core/spec.md

  "labels": { "STRING": "STRING" * }, ?

  "model": {                            # Only if inlined
    "schemas": [ "STRING" * ], ?        # Available schema formats
    "attributes": {                     # Registry level extensions
      "STRING": {                       # Attribute name (case sensitive)
        "name": "STRING",               # Same as attribute's key
        "type": "TYPE",                 # string, decimal, array, object, ...
        "description": "STRING", ?
        "enum": [ VALUE * ], ?          # Array of scalar values of type "TYPE"
        "strict": BOOL, ?               # Just "enum" values or not. Def=true
        "required": BOOL, ?             # Default: false, from a CLI POV?
        "attributes": { ... }, ?        # Only if "type" == "object"
        "keyType": "TYPE", ?            # Only if "type" == "map"
        "itemType": "TYPE", ?           # Only if "type" == "array" or "map"

        "ifValue": {                    # Only if "type" != "object"
          VALUE: {                      # Possible attribute value
            "siblingAttributes": { ... } # See "attributes" above
          } *
        } ?
      } *
    },

    "groups": {

>       "STRING": {                       # Plural name, eg. "endpoints"

Reviewer: Should this be on the next line instead?

Author: I wanted to make it clear that they key was the plural name, not the singular one.
I'll see `key=` to the from of the comment to make that clearer.

Reviewer: I think that would be clearer, yes. My confusion is that the name line really *is* the plural name. (Is it an error if the key doesn't match the value of `plural`? And do we note that?)

Author: Good great - I forgot to change it from Array to Map and I added the MUST for both Groups and Resources:
```
-  - Type: Array
+  - Type: Map where the key MUST be the plural name of the Group
```

        "plural": "STRING",             # eg. "endpoints"

>         "singular": "STRING",           # eg. "endpoint"

Reviewer: If we're editing this a lot anyway, let's correct "eg." to "e.g." everywhere :)

Author: was trying to save space but let me see if anything wraps.... ok we're good.

        "attributes": { ... }, ?        # See "attributes" above

        "resources": {

>           "STRING": {                   # Plural name, e.g. "definitions"

Reviewer: Again, I think this should be in the next line.

            "plural": "STRING",         # eg. "definitions"
            "singular": "STRING",       # eg. "definition"

>             "versions": UINT ?          # Num Versions(>=0). Def=1, 0=unlimited

Reviewer: What does "Def" mean here?

Author: "Default" - I added the "ault" w/o wrapping :-)

            "versionId": BOOL, ?        # Supports client specified Version IDs
            "latest": BOOL, ?           # Supports client "latest" selection
            "hasDocument": BOOL, ?      # Has a separate document. Def=true
            "attributes": { ... } ?     # See "attributes" above
          } *
        } ?
      } *
    } ?
  } ?

  # Repeat for each Group type
  "GROUPsUrl": "URL",                              # eg. "endpointsUrl"
  "GROUPsCount": INT                               # eg. "endpointsCount"
  "GROUPs": {                                      # Only if inlined

>     "ID": {                                        # The Group id

Reviewer: This comment and the next line look like they're the wrong way round. Or at least odd.

### Attributes and Extensions

Unless otherwise noted, all attributes and extensions MUST be mutable and MUST
be one of the data types defined in the
[Notational Conventions](#notational-conventions) section. In other words, one
of: array, bool, decimal, map, object, string, time, uint, uri, uri-reference,
uri-tempalte, url.

Implementations of this specification MAY define additional (extension)
attributes. However they MUST adhere to the following rules:

- they MUST be defined as part of the [Registry Model](#registry-model)
- the presence of an undefined attribute in a request MUST generate an error
- they MUST NOT use the name of an attribute defined by this specification,
  regardless of which entity the attribute is defined for. In other words,
  a spec defined attribute for GROUPs can not be reused as an extension for
  a RESOURCE
- it is RECOMMENDED that extension attributes on different objects not use the

>   same name unless they have the exact same semantics meaning

Reviewer: ```suggestion
  same name unless they have the exact same semantic meaning
```

``` text
{
  "schemas": [ "STRING" * ], ?         # Available schema formats
  "attributes": {                      # Registry level extensions
    "STRING": {                        # Attribute name
      "name": "STRING",                # Same as attribute's key
      "type": "TYPE",                  # bool, string, array, object, ...
      "description": "STRING",
      "enum": [ VALUE * ], ?           # Array of values of type "TYPE"
      "strict": BOOL, ?                # Just "enum" values or not. Def=true
      "required": BOOL, ?              # Default: false
      "attributes": { ... }, ?         # Only if "type" == "object". Children
      "keyType": "TYPE", ?             # Only if "type" == "map"
      "itemType": "TYPE", ?            # Only if "type" == "array"

      "ifValue": {                     # Conditional extensions
        VALUE: {
          "siblingAttributes": { ... } # Siblings to this "attribute"
        } *
      }
    } *
  },
  "groups": {

>     "STRING": {                        # Plural name, eg. "endpoints"

Reviewer: Again with the slightly odd location of comments. I'm beginning to think I just don't understand.

Reviewer: Massively, massively - it's somewhat shocking just how much clearer that makes it, for me anyway :)

TODO: check the above example list

- `attributes."STRING".name
  - The name of the attribute. MUST be the same as the key used in the owning
    `attributes` attribute.

>   - Type: Sring

Reviewer: ```suggestion
  - Type: String
```

Author: done

`EXPRESSION` does not match the entity, that entity MUST NOT be returned. In
other words, a `404 Not Found` would be generated in the HTTP protocol case.

At this time, this specification only supports filtering over scalar attributes
defined at the root of the xRegistry entities (the Registry itself, Groups,

> Resources and Versions). The one exception to this are maps (such as `labels`)

Reviewer: The singular/plural clash here is bugging me.
```suggestion
Resources and Versions). The one exception to this is support for filtering on maps (such as `labels`)
```

Author: done
